lib/outdated.js

```diff
@@ -1,437 +1,273 @@
-    seen.add(item)
-    uniqed.push(item)
-  })
-  return uniqed
+const usageUtil = require('./utils/usage.js')
+const ansiTrim = require('./utils/ansi-trim.js')
+
+const usage = usageUtil('outdated',
+  'npm outdated [[<@scope>/]<pkg> ...]'
+)
+const completion = (cb) => cb(null, [])
+
+function cmd (args, cb) {
+  outdated(args, cb)
+    .then(() => cb())
+    .catch(cb)
 }
 
-function andComputeMetadata (next) {
-  return function (er, tree) {
-    if (er) return next(er)
-    next(null, computeMetadata(tree))
-  }
-}
+async function outdated (args) {
+  const opts = npm.flatOptions
+  const global = path.resolve(npm.globalDir, '..')
+  const where = opts.global
+    ? global
+    : npm.prefix
 
-function outdated (args, silent, cb) {
-  if (typeof cb !== 'function') {
-    cb = silent
-    silent = false
-  }
-  let opts = npm.flatOptions
-  var dir = path.resolve(npm.dir, '..')
-
-  // default depth for `outdated` is 0 (cf. `ls`)
-  if (opts.depth === Infinity) opts = { ...opts, depth: 0 }
-
-  readPackageTree(dir, andComputeMetadata(function (er, tree) {
-    if (!tree) return cb(er)
-    mutateIntoLogicalTree(tree)
-    outdated_(args, '', tree, {}, 0, opts, function (er, list) {
-      list = uniq(list || []).sort(function (aa, bb) {
-        return aa[0].path.localeCompare(bb[0].path) ||
-          aa[1].localeCompare(bb[1])
-      })
-      if (er || silent ||
-          (list.length === 0 && !opts.json)) {
-        return cb(er, list)
-      }
-      if (opts.json) {
-        output(makeJSON(list, opts))
-      } else if (opts.parseable) {
-        output(makeParseable(list, opts))
-      } else {
-        var outList = list.map(x => makePretty(x, opts))
-        var outHead = [ 'Package',
-          'Current',
-          'Wanted',
-          'Latest',
-          'Location'
-        ]
-        if (long) outHead.push('Package Type', 'Homepage')
-        var outTable = [outHead].concat(outList)
-
-        if (opts.color) {
-          outTable[0] = outTable[0].map(function (heading) {
-            return styles.underline(heading)
-          })
-        }
-
-        var tableOpts = {
-          align: ['l', 'r', 'r', 'r', 'l'],
-          stringLength: function (s) { return ansiTrim(s).length }
-        }
-        output(table(outTable, tableOpts))
-      }
-      process.exitCode = list.length ? 1 : 0
-      cb(null, list.map(function (item) { return [item[0].parent.path].concat(item.slice(1, 7)) }))
-    })
-  }))
-}
+  const arb = new Arborist({
+    ...opts,
+    path: where
+  })
 
-// [[ dir, dep, has, want, latest, type ]]
-function makePretty (p, opts) {
-  var depname = p[1]
-  var has = p[2]
-  var want = p[3]
-  var latest = p[4]
-  var type = p[6]
-  var deppath = p[7]
-  var homepage = p[0].package.homepage || ''
-
-  var columns = [ depname,
-    has || 'MISSING',
-    want,
-    latest,
-    deppath || 'global'
-  ]
-  if (long) {
-    columns[5] = type
-    columns[6] = homepage
-  }
+  const tree = await arb.loadActual()
```

Comment:
I was thinking we might want this to be `arb.buildIdealTree()` so it finds outdated deps in the virtual tree, falling back to the actual if there's no lockfile.  But I wonder if that even makes sense?  The current behavior only looks at the actual tree, but it also predates the existence of lockfiles, so I'm not sure how much sense it makes to go based on that.